hazelcast-sql-core/src/main/java/com/hazelcast/sql/impl/calcite/validate/operators/CoalesceFunction.java

```diff
@@ -0,0 +1,147 @@
+
+import com.hazelcast.sql.impl.QueryException;
+import com.hazelcast.sql.impl.SqlErrorCode;
+import com.hazelcast.sql.impl.calcite.validate.HazelcastCallBinding;
+import com.hazelcast.sql.impl.calcite.validate.HazelcastSqlValidator;
+import com.hazelcast.sql.impl.calcite.validate.operators.common.HazelcastFunction;
+import com.hazelcast.sql.impl.calcite.validate.types.HazelcastTypeUtils;
+import com.hazelcast.sql.impl.type.QueryDataType;
+import com.hazelcast.sql.impl.type.QueryDataTypeFamily;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.sql.SqlCall;
+import org.apache.calcite.sql.SqlFunctionCategory;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.SqlLiteral;
+import org.apache.calcite.sql.SqlNode;
+import org.apache.calcite.sql.SqlOperandCountRange;
+import org.apache.calcite.sql.SqlUtil;
+import org.apache.calcite.sql.type.ReturnTypes;
+import org.apache.calcite.sql.type.SqlOperandCountRanges;
+import org.apache.calcite.sql.validate.SqlValidatorScope;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static com.hazelcast.sql.impl.calcite.validate.operators.HazelcastReturnTypeInference.wrap;
+import static org.apache.calcite.util.Static.RESOURCE;
+
+public final class CoalesceFunction extends HazelcastFunction {
+    public static final CoalesceFunction INSTANCE = new CoalesceFunction();
+
+    private CoalesceFunction() {
+        super(
+                "COALESCE",
+                SqlKind.COALESCE,
+                wrap(ReturnTypes.ARG0_NULLABLE),
+                VariableLengthOperandTypeInference.INSTANCE,
+                SqlFunctionCategory.SYSTEM);
+    }
+
+    @Override
+    public SqlOperandCountRange getOperandCountRange() {
+        return SqlOperandCountRanges.any();
+    }
+
+    @Override
+    protected boolean checkOperandTypes(HazelcastCallBinding callBinding, boolean throwOnFailure) {
+        HazelcastSqlValidator validator = callBinding.getValidator();
+        SqlValidatorScope scope = callBinding.getScope();
+
+        SqlCall call = callBinding.getCall();
+        List<SqlNode> operandList = call.getOperandList();
+
+        List<RelDataType> argTypes = new ArrayList<>(operandList.size());
+
+        boolean foundNotNull = false;
+        for (SqlNode node : operandList) {
+            argTypes.add(validator.deriveType(scope, node));
+            foundNotNull |= !SqlUtil.isNullLiteral(node, false);
+        }
+
+        if (!foundNotNull) {
+            if (throwOnFailure) {
+                throw callBinding.newError(RESOURCE.mustNotNullInElse());
+            }
+            return false;
+        }
+
+        RelDataType operandType = argTypes.get(0);
+        for (int i = 1; i < argTypes.size(); i++) {
+            operandType = HazelcastTypeUtils.withHigherPrecedence(operandType, argTypes.get(i));
+        }
+
+        QueryDataType caseHzReturnType = HazelcastTypeUtils.toHazelcastType(operandType.getSqlTypeName());
```

Comment:
I guess `caseHzReturnType` could be derived from `operandType` inside `allBranchTypesCanBeConvertedToReturnType()`.